[ALLUXIO-1893] Support deploying alluxio on secure yarn cluster

lshmouse opened the pull request:
As title~

AmplabJenkins commented:
Can one of the admins verify this patch?

gpang commented:
ok to test

gpang commented:
Jenkins, add to whitelist

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9378/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-integration-yarn' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9378/org.alluxio$alluxio-integration-yarn/testReport'>org.alluxio:alluxio-integration-yarn</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9378/org.alluxio$alluxio-integration-yarn/testReport/org.apache.maven.surefire.junit4/JUnit4Provider/alluxio_yarn_ApplicationMasterTest/'><strong>org.apache.maven.surefire.junit4.JUnit4Provider.alluxio.yarn.ApplicationMasterTest</strong></a></li></ul>
Test FAILed.

gpang commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9379/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-integration-yarn' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9379/org.alluxio$alluxio-integration-yarn/testReport'>org.alluxio:alluxio-integration-yarn</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9379/org.alluxio$alluxio-integration-yarn/testReport/org.apache.maven.surefire.junit4/JUnit4Provider/alluxio_yarn_ApplicationMasterTest/'><strong>org.apache.maven.surefire.junit4.JUnit4Provider.alluxio.yarn.ApplicationMasterTest</strong></a></li></ul>
Test FAILed.

lshmouse commented:
I think the failed test has no relation with this patch.

I have run the tests for several times and it passed every times

```
Running alluxio.yarn.ApplicationMasterTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.889 sec
Results :
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0
```

lshmouse commented:
@gpang Could you help to push this request?

apc999 commented on integration/yarn/src/main/java/alluxio/yarn/YarnUtils.java:
```diff
@@ -147,5 +149,17 @@ public static String buildCommand(YarnContainerType containerType, Map<String, S
+   */
+  public static InetSocketAddress getRMAddress(Configuration config) {
+    return config.getSocketAddr(YarnConfiguration.RM_ADDRESS,
+            YarnConfiguration.DEFAULT_RM_ADDRESS,
```
NIT: the indent doesn't look correct to me (should be 4 space on line breaks), please checkout http://www.alluxio.org/documentation/v1.0.0/en/Contributing-to-Alluxio.html#coding-style

apc999 commented on integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:
```diff
@@ -238,6 +264,21 @@ public float getProgress() {
   public void start() throws IOException, YarnException {
+    if (UserGroupInformation.isSecurityEnabled()) {
+      Credentials credentials =
+              UserGroupInformation.getCurrentUser().getCredentials();
```
indent is off here.

apc999 commented on integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:
```diff
@@ -175,17 +185,33 @@ public static void main(String[] args) {
+          String resourcePath = cliParser.getOptionValue("resource_path");
+
+          ApplicationMaster applicationMaster =
+                  new ApplicationMaster(numWorkers, masterAddress, resourcePath);
```
indent is off here.

lshmouse commented:
Thanks for @apc999 's review! 
The indent problems have been fixed in the new patch.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9439/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-integration-yarn' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9439/org.alluxio$alluxio-integration-yarn/testReport'>org.alluxio:alluxio-integration-yarn</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9439/org.alluxio$alluxio-integration-yarn/testReport/org.apache.maven.surefire.junit4/JUnit4Provider/alluxio_yarn_ApplicationMasterTest/'><strong>org.apache.maven.surefire.junit4.JUnit4Provider.alluxio.yarn.ApplicationMasterTest</strong></a></li></ul>
Test FAILed.

apc999 commented:
@lshmouse thanks for the update.
It seems the unit test of application master keeps failing,
```
java.lang.RuntimeException: test should never throw an exception to this level
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:100)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.executeTest(PowerMockJUnit44RunnerDelegateImpl.java:294)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTestInSuper(PowerMockJUnit47RunnerDelegateImpl.java:127)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTest(PowerMockJUnit47RunnerDelegateImpl.java:82)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runBeforesThenTestThenAfters(PowerMockJUnit44RunnerDelegateImpl.java:282)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:50)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.invokeTestMethod(PowerMockJUnit44RunnerDelegateImpl.java:207)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.runMethods(PowerMockJUnit44RunnerDelegateImpl.java:146)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$1.run(PowerMockJUnit44RunnerDelegateImpl.java:120)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:34)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:44)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.run(PowerMockJUnit44RunnerDelegateImpl.java:122)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.run(JUnit4TestSuiteChunkerImpl.java:106)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.run(AbstractCommonPowerMockRunner.java:53)
	at org.powermock.modules.junit4.PowerMockRunner.run(PowerMockRunner.java:59)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
```

Do you have any idea why it is failing?

lshmouse commented:
@apc999 
I run the test in my local for many times and can not reproduce the failure.

From the code: http://grepcode.com/file/repo1.maven.org/maven2/junit/junit/4.8.1/org/junit/internal/runners/MethodRoadie.java#97

MethodRoadie.java catch an Exception, but don't print it out.

Any suggestion? Thanks~

aaudiber commented:
@lshmouse I get the exception 100% of the time when I run the test through maven on this branch. The following command fails. Is this what you're running?
```
mvn test -Pdeveloper -Dtest='ApplicationMasterTest' -Dmaven.javadoc.skip -Dfindbugs.skip -Dcheckstyle.skip -Dlicense.skip -DfailIfNoTests=false
```

From https://github.com/jayway/powermock/issues/481, it sounds like this is probably caused by the security code throwing an exception which doesn't have a stack trace. Do you know why this might be happening?

lshmouse commented:
@aaudiber 
Thanks for your advice. I reproduced the failure and fixed it. Thanks~

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9472/<h2>Build result: FAILURE</span></h2>[...truncated 3015 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/client/target/alluxio-keyvalue-client-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-client/1.1.0-SNAPSHOT/alluxio-keyvalue-client-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/pom.xml to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/hdfs/target/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-hdfs/1.1.0-SNAPSHOT/alluxio-underfs-hdfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/pom.xml to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/swift/target/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-swift/1.1.0-SNAPSHOT/alluxio-underfs-swift-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.1.0-SNAPSHOT/alluxio-examples-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/pom.xml to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/target/alluxio-core-server-1.1.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-server/1.1.0-SNAPSHOT/alluxio-core-server-1.1.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/pom.xml to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.1.0-SNAPSHOT/alluxio-keyvalue-common-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/pom.xml to org.alluxio/alluxio-underfs/1.1.0-SNAPSHOT/alluxio-underfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/underfs/glusterfs/target/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-underfs-glusterfs/1.1.0-SNAPSHOT/alluxio-underfs-glusterfs-1.1.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/pom.xml to org.alluxio/alluxio-integration/1.1.0-SNAPSHOT/alluxio-integration-1.1.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/pom.xml to org.alluxio/alluxio-keyvalue/1.1.0-SNAPSHOT/alluxio-keyvalue-1.1.0-SNAPSHOT.pomchannel stopped
Test FAILed.

aaudiber commented:
Looks like there's a minor checkstyle violation:
```
/home/jenkins/workspace/Alluxio-Pull-Request-Builder/integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:215: error: First sentence should end with a period.
```

apc999 commented on integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:
```diff
@@ -125,6 +133,8 @@
    */
   private CountDownLatch mOutstandingWorkerContainerRequestsLatch = null;
 
+  private ByteBuffer mAllTokens;
```
NIT: could you brief javadoc this member variable?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9486/
Test PASSed.

lshmouse commented:
@apc999  @aaudiber 
Could you help to push this issue?

apc999 commented:
@lshmouse thanks for the update. the PR looks good to me.

lshmouse commented:
@gpang
Could you help to merge this request?

apc999 commented:
@gpang @aaudiber , could you take a look at this PR? It has been open for a while

aaudiber commented on integration/yarn/src/test/java/alluxio/yarn/ApplicationMasterTest.java:
```diff
@@ -109,6 +110,7 @@
       ImmutableMap.<String, String>builder()
           .put("CLASSPATH", ENV_CLASSPATH)
           .put("ALLUXIO_HOME", ApplicationConstants.Environment.PWD.$())
+          .put("ALLUXIO_RAM_FOLDER", ApplicationConstants.Environment.LOCAL_DIRS.$())
```
Is the idea here to mount the ramdisk inside the Yarn container so that Yarn cleans it up on exit?

lshmouse replied:
Yes

aaudiber commented on integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:
```diff
@@ -124,6 +132,8 @@
    * {@link #launchAlluxioWorkerContainers(List)}.
    */
   private CountDownLatch mOutstandingWorkerContainerRequestsLatch = null;
+  /** Security tokens for HDFS.*/
```
nit: space after period

lshmouse replied:
done

aaudiber commented on integration/yarn/src/main/java/alluxio/yarn/ApplicationMaster.java:
```diff
@@ -238,6 +276,21 @@ public float getProgress() {
+    if (UserGroupInformation.isSecurityEnabled()) {
+      Credentials credentials =
+          UserGroupInformation.getCurrentUser().getCredentials();
+      DataOutputBuffer dob = new DataOutputBuffer();
```
nit: maybe rename this to credentialsBuffer?

lshmouse replied:
done

aaudiber commented:
@lshmouse There are some merge conflicts with import statements, but after resolving those I was able to test this and it works great! I left a couple minor comments on the code, but otherwise this looks good to me. Thanks for the contribution!

aaudiber commented:
@lshmouse Feel free to reopen if you have time to iterate on this - otherwise closing for now due to inactivity

lshmouse commented:
Reopen this PR and rebase the patch on master

lshmouse commented:
@aaudiber  @apc999 
I rebased the PR on master and updated it for @aaudiber 's review. 
But it seems that I can't reopen the PR.  Should I send a new One?

aaudiber commented:
@lshmouse Sorry, it looks like the branch got deleted so we can't reopen. Could you create a new PR?

lshmouse commented:
OK. Thanks @aaudiber.
A new PR: #3806 has been sent~
